### PR TITLE
[C#] Expose Multi-Lora support in C#

### PR DESCRIPTION
**Pull request by yuslepukhin**
### Motivation and Context
https://github.com/microsoft/onnxruntime/pull/22046

### CODE REVIEW

File: csharp/src/Microsoft.ML.OnnxRuntime/OrtLoraAdapter.shared.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            var platformPath = NativeOnnxValueHelper.GetPlatformSerializedString(adapterPath);
+            var allocatorHandle = (ortAllocator != null) ? ortAllocator.Pointer : IntPtr.Zero;
+            NativeApiStatus.VerifySuccess(NativeMethods.CreateLoraAdapter(platformPath, allocatorHandle,
+                out IntPtr adapterHandle));
+            return new OrtLoraAdapter(adapterHandle);
+        }
+
+        /// <summary>
+        /// Creates an instance of OrtLoraAdapter from an array of bytes.
```
Comment (skottmckay) [inline csharp/src/Microsoft.ML.OnnxRuntime/OrtLoraAdapter.shared.cs:32]: Would be good to point out that the bytes are copied into the new OrtLoraAdapter so the user knows they don't need to remain valid.

File: csharp/src/Microsoft.ML.OnnxRuntime/RunOptions.shared.cs [hunk 2]
```diff
@@ -129,6 +129,17 @@ public void AddRunConfigEntry(string configKey, string configValue)
             NativeApiStatus.VerifySuccess(NativeMethods.OrtAddRunConfigEntry(handle, utf8Key, utf8Value));
         }
 
+        /// <summary>
+        /// Appends the specified lora adapter to the list of active lora adapters
+        /// for this RunOptions instance. All run calls with this instant will
+        /// make use of the activates Lora Adapters
+        /// </summary>
+        /// <param name="loraAdapter">Lora adapter instance</param>
+        public void AddActiveLoraAdapter(OrtLoraAdapter loraAdapter)
```
Comment (skottmckay) [inline csharp/src/Microsoft.ML.OnnxRuntime/RunOptions.shared.cs:138]: Is there a notion of an 'inactive' LoRA adapter? Given it's in RunOptions I would have though the 'active' was implied.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs [hunk 3]
```diff
@@ -1251,6 +1251,87 @@ private void TestModelSerialization()
             }
         }
 
+        private static OrtLoraAdapter CreateLoraAdapterFromFile()
+        {
+            var adapterPath = Path.Combine(Directory.GetCurrentDirectory(), "two_params_lora_model.onnx_adapter");
```
Comment (skottmckay) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs:1256]: Out of interest is the formatting for the ORT LoRA adapter documented somewhere? ideally with an example of how to create/use one ?

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs [hunk 4]
```diff
@@ -1251,6 +1251,87 @@ private void TestModelSerialization()
...
+            var inputShape = new long[] { 4, 4 };
+            var inputData = new float[16];
+            Array.Fill(inputData, 1);
+            using var inputOrtValue = OrtValue.CreateTensorValueFromMemory(inputData, inputShape);
+
+            var expectedOutput = new float[] {
+                  154, 176, 198, 220,
+                  154, 176, 198, 220,
+                  154, 176, 198, 220,
+                  154, 176, 198, 220 };
```
Comment (skottmckay) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs:1281]: nit: it's a little confusing to have separate tests for the with/without LoRA adapter being used. 

Can we combine them given (IIUC) we should be able to re-use the InferenceSession and just change the RunOptions between the execution w/ and w/o the adapter. 

or if they're separate can the base test without the adapter be first, and there be a comment in `testInferenceWithLoraAdapter` saying to see `testInferenceWithBaseLoraModel` for the base output?

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs (hunk 3 above)
Comment (yuslepukhin) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs:1256]: Yes, it is documented at onnxruntime/lora/adapter_format

**Review by skottmckay — APPROVED** [review]

**Pull request merged**